source/class/qx/Class.js

```diff
@@ -305,1400 +657,2442 @@ qx.Bootstrap.define("qx.Class", {
+            return v !== null && fontManager.isDynamic(v);
+          }
+        ]
+      ]);
+
+      // If the property is nullable and the value is null...
+      if (property.nullable && value === null) {
+        // ... then we don't do the check
+      } else if (property.inheritable && value == "inherit") {
+        // Request to explicitly inherit from
+        // parent. Ignore check.
+      } else if ($$checks.has(property.check)) {
+        if (!$$checks.get(property.check)(value)) {
           throw new Error(
-            'Invalid type "' +
-              config.type +
-              '" definition for class "' +
-              name +
-              '"!'
+            `${prop}: ` +
+              `Expected value to be of type ${property.check}; ` +
+              `value=${value}`
           );
         }
-
-        // Validate non-static class on the "extend" key
-        if (config.type && config.type !== "static" && !config.extend) {
+      } else if (typeof property.check == "function") {
+        if (!property.check.call(proxy, value)) {
           throw new Error(
-            'Invalid config in class "' +
-              name +
-              '"! Every non-static class has to extend at least the "qx.core.Object" class.'
+            `${prop}: ` +
+              `Check function indicates wrong type value; ` +
+              `value=${value}`
           );
         }
+      } else if (Array.isArray(property.check)) {
+        if (value instanceof String) {
+          value = value.valueOf();
+        }
 
-        // Validate keys
-        var allowed =
-          config.type === "static"
-            ? this.__staticAllowedKeys
-            : this.__allowedKeys;
-        for (var key in config) {
-          if (!allowed[key]) {
-            throw new Error(
-              'The configuration key "' +
-                key +
-                '" in class "' +
-                name +
-                '" is not allowed!'
-            );
-          }
+        qx.core.Assert.assertInArray(
+          value,
+          property.check,
+          "Expected value to be one of: [" + property.check + "]"
+        );
+      } else if (typeof property.check == "string") {
+        if (qx.Class.isDefined(property.check)) {
+          qx.core.Assert.assertInstance(
+            value,
+            qx.Class.getByName(property.check),
+            "Expected value to be an instance of " + property.check
+          );
+        } else if (qx.Interface && qx.Interface.isDefined(property.check)) {
+          qx.core.Assert.assertInterface(
+            value,
+            qx.Interface.getByName(property.check),
+            "Expected value to implement " + property.check
+          );
+        } else {
+          if (qx.core.Environment.get("qx.Class.futureCheckJsDoc")) {
+            // Next  try to parse the check string as JSDoc
+            let bJSDocParsed = false;
+            try {
+              const { parse } = require("jsdoctypeparser");
```

Comment:
I appreciate that this is currently switched off, but IMHO we should not plan on including require in browser based builds because it forces the user to use npm and have a package; years ago, one of the 1&1 team added grunt and it was very frustrating to have to create a package.json etc just to be able to compile.

Reply:
Remember that `require` in a browser-based app includes the browserified code. There *should* be o need for the user to use npm and have a package; the compiler should be browserifying that code from qooxdoo's node_modules.

That said, I don't have any strong need for this feature. We could strip it out completely. We could, later, implement it with our own code to parse JSDoc comments. Or whatever. Let me know if you'd like for me to entirely remove that code.

Reply:
I think that the jsdoc parser would be useful, but at the moment browserify does not work as you're propising (although I think it would be good if it did).  I havn't specifically tested, but the compiler was complaining that it could not find the module

Reply:
> I think that the jsdoc parser would be useful, but at the moment browserify does not work as you're propising (although I think it would be good if it did). I havn't specifically tested, but the compiler was complaining that it could not find the module

I just pushed a fix to this. The compiler now looks in both the app's node_modules directory and qooxdoo's node_modules directory for browserifyable modules. This means that `jsdoctypeparser`, used by `Class.js` (when enabled by the environment variable, later), is found regardless of whether there is a node_modules directory in the user's app.